src/frequenz/sdk/timeseries/_ringbuffer.py

```diff
@@ -0,0 +1,430 @@
+                "2022-01-01 12:00:00" to "2022-01-02 12:00:00" (date chosen
+                arbitrarily here).
+        """
+        self._buffer = buffer
+        self._sampling_period = sampling_period
+        self._time_index_alignment = time_index_alignment
+
+        self._gaps: list[Gap] = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+        self._time_range = (len(self._buffer) - 1) * sampling_period
+
+    @property
+    def sampling_period(self) -> timedelta:
+        """Return the sampling period of the ring buffer.
+
+        Returns:
+            Sampling period of the ring buffer.
+        """
+        return self._sampling_period
+
+    @property
+    def gaps(self) -> List[Gap]:
+        """Get the list of ranges for which no values were provided.
+
+        See definition of dataclass @Gaps for more info.
+
+        Returns:
+            List of gaps.
+        """
+        return self._gaps
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            The max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> None:
+        """Update the buffer with a new value for the given timestamp.
+
+        Args:
+            timestamp: Timestamp of the new value.
+            value: value to add.
+            missing: if true, the given timestamp will be recorded as missing.
+                The value will still be written.
+
+                Note: When relying on your own values to mark missing data (e.g.
+                `math.nan`), be advised that when `update()` is called with
+                timestamps newer than the current time + `sampling_period`
+                (as is the case for loading historical data after an app
+                restart), a gap of missing data exists.
+                This gap does not contain NaN values but simply the old invalid values.
+                The list of gaps returned by `gaps()` will reflect this.
+
+        Raises:
+            IndexError: When the timestamp to be added is too old.
+        """
+        # adjust timestamp to be exactly on the sample period time point
+        timestamp = self._normalize_timestamp(timestamp)
+
+        # Don't add outdated entries
+        if timestamp < self._datetime_oldest and self._datetime_oldest != datetime.max:
+            raise IndexError(
+                f"Timestamp too old (cut-off is at {self._datetime_oldest})."
+            )
+
+        # Update timestamps
+        prev_newest = self._datetime_newest
+        self._datetime_newest = max(self._datetime_newest, timestamp)
+        self._datetime_oldest = self._datetime_newest - self._time_range
+
+        # Update data
+        self._buffer[self.datetime_to_index(timestamp)] = value
+
+        self._update_gaps(timestamp, prev_newest, missing)
+
+    def datetime_to_index(
```

Comment:
In python this is also commonly known as simply `index()` (for example `list` implements it).

Reply:
Do you think we should rename it? :)